Pytree refactor

billbrod opened the pull request:
Just opening this right now so that we can get the GLM example rendered by read the docs. More work needs to be done here.

For the example in particular:
- I'm using a dataset from DANDI, looking at head direction and 2d position.
- I think this example is too complex to introduce the pytrees and more/better text is needed here.
- Writing this made it clear that we can easily put together some simple helper functions that will make things easier, like doing the multiplication between bases and coefficients.
- When this goes into the docs, the general "here's a pytree" should probably be a separate tutorial from "here's how to use them with GLMs".
- I added some extra dependencies in pyproject.toml, but that's just so this example builds. We should think about how we want to handle this.

gviejo commented on docs/examples/plot_glm_pytree.py:
```diff
@@ -0,0 +1,320 @@
+spikes = spikes.count(bin_size=.02, time_units='s').restrict(valid_data)
+
+# add extra dim for n_neurons (here, 1)
+X = nmo.pytrees.FeaturePytree(head_direction=np.expand_dims(basis.evaluate(head_dir.values), 1))
```
Also why the expand_dims?

billbrod replied:
as discussed, expand_dims adds an extra dimension for (the single) neurons. We should think about how to handle this, but it's not a pytree problem

billbrod commented:
I addressed Guillaume's comments and added tests for pytrees. There weren't many, but I think they're sufficient. Let me know if there's anything else I should add to them.

billbrod commented:
Currently, one test is failing:
- `test_fit_init_params_type`: when input is `iter` or a dict with keys `p1,p2`. The problem is that we expect `params` to be a 2-tuple (in order  to be compliant with sklearn, I thin), but this leads to some weirdness, as in these cases. We should talk through this

BalzaniEdoardo commented on docs/examples/plot_glm_pytree.py:
```diff
@@ -0,0 +1,320 @@
+mapped['feature_1']
+# %%
+# We can change the dimensionality of our pytree:
+mapped = jax.tree_map(lambda x: jnp.mean(x, axis=-1), example_pytree)
```
Agree

BalzaniEdoardo reviewed (changes requested):
great work; I added a couple of commits. 
**First commit content:**
1. Add 1-line docstrings to tests.
2. Assert that tree_map returns a FeaturePytree
3. moved the --cov from the general config to the tox.ini call to pytest; the debugger behaves poorly if tests are configured with cov (doesn't stop at breakpoints)

**Second commit content:**
1. Added a test for asserting the equivalence of fitting a glm using pytrees or arrays

BalzaniEdoardo reviewed (approved):
good to go!
